src/compiler/crystal/tools/formatter.cr

```diff
@@ -4775,17 +4775,26 @@ module Crystal
       result
     end
 
-    def fix_heredocs(lines, @heredoc_fixes)
+    def fix_heredocs(lines)
       @heredoc_fixes.each do |fix|
+        min_difference = (fix.start_line..fix.end_line).min_of do |line_number|
+          leading_space_count(lines[line_number], fix.difference)
+        end
+
         fix.start_line.upto(fix.end_line) do |line_number|
           line = lines[line_number]
-          if (0...fix.difference).all? { |index| line[index]?.try &.ascii_whitespace? }
-            lines[line_number] = line[fix.difference..-1]
-          end
+          lines[line_number] = line[min_difference..]
         end
       end
     end
 
+    def leading_space_count(line, max_count)
+      max_count.times do |index|
+        return index unless line[index]?.try &.ascii_whitespace?
```

Comment:
You could use an unsafe byte accessor here because ASCII whitespace values at the start are all only a single byte wide.